### PR TITLE
[9.0][MIG] web_widget_image_download

### PR DESCRIPTION
**Pull request by kutyel**
# Web Widget - Image Download

This module was written to extend the functionality of the widget image and allows to download it.
# Usage

To use this module, you need to:
1. Go to the section `Contacts`.
2. Click on a contact.
3. Edit the contact.
4. Click the `Download` button (between `Edit` and `Clear`).

@Tecnativa

### CODE REVIEW

File: web_widget_image_download/static/src/js/web_widget_image_download.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2016 Flavio Corpa <flavio.corpa@tecnativa.com>
+ * License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+ */
+odoo.define('web_widget_image_download', function (require) {
```
Comment (yajo) [inline web_widget_image_download/static/src/js/web_widget_image_download.js]: Module name should follow [the convention `module_name.servie`.](https://www.odoo.com/documentation/9.0/reference/javascript.html#javascript-module-system-overview)

File: web_widget_image_download/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Flavio Corpa <flavio.corpa@tecnativa.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
```
Comment (yajo) [inline web_widget_image_download/__init__.py]: No need to copyright an empty file..

File: web_widget_image_download/static/src/js/web_widget_image_download.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            // use jquery instead of `replace` with qweb (to avoid breaking inheritance)
+            if (this.has_custom_image()) {
+                this.$el.find('.oe_form_binary_file_clear').removeClass('col-md-offset-5');
+            }
+
+            $widget.attr('href', this.imgSrc);
+        },
+        has_custom_image: function () {
+            // check if the image of the widget is different from the default placeholder
+            return this.imgSrc && !this.imgSrc.includes('/placeholder.png');
```
Comment (yajo) [inline web_widget_image_download/static/src/js/web_widget_image_download.js]: You can use `this.imgSrc == this.placeholder`

File: web_widget_image_download/static/src/js/web_widget_image_download.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2016 Flavio Corpa <flavio.corpa@tecnativa.com>
+ * License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+ */
+odoo.define('web_widget_image_download', function (require) {
+    'use strict';
+
+    var common = require('web.form_common');
```
Comment (yajo) [inline web_widget_image_download/static/src/js/web_widget_image_download.js]: Maybe the problem is that this should require `web.form_widgets` instead?

**Comment by yajo** [comment]
I fixed this myself, it should be ready to merge.

**Comment by kutyel** [comment]
Looks perfect, thanks! :+1:

**Comment by carlosdauden** [comment]
:+1:

**Comment by pedrobaeza** [comment]
Going to Setting > Companies > Companies, I'm not able to download the company image. I click on the button but nothing happens.

File: web_widget_image_download/README.rst [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/162/8.0
+
+Known Issues / Roadmap
+======================
+
+* In order to work correctly, this widget has to detect image type, the server should include this information in the `Content-Type` header. Right now, odoo is not doing so, but a fix has been `proposed <https://github.com/odoo/odoo/pull/12918>`_.
+* For some unknown reason, the widget does not work in the `Preferences` view, because odoo is not rendering the **QWeb** template.
```
Comment (kutyel) [inline web_widget_image_download/README.rst]: @Yajo you should remove this from the **Known Issues**, as it is solved by your last commit (https://github.com/OCA/web/pull/392/commits/68b8b03b3b462b353f41600043396094e47d69a2), the same happens in the PR for the v8.0. 👍

**Comment by yajo** [comment]
Linking this to https://github.com/odoo/odoo/pull/12918 and https://github.com/OCA/OCB/issues/523. It will fail functionally outside of OCB.

**Comment by yajo** [comment]
I hope that the point of me being the fixer of my own comments doesn't invalidate my thumb: :+1:

**Comment by lasley** [comment]
👍

**Comment by pedrobaeza** [comment]
:+1:

**Pull request merged**